File: binsifter/core/yara_rule_gen.py
"""Draft YARA rule auto-generation per SSDEEP cluster - direct port of the
v1.3-proto1 PowerShell logic (BinSifter-Rowan.ps1, lines
~3225-3320).

Best-effort, clearly-labeled-as-draft rules built from strings common to
every member of a size>=2 SSDEEP cluster. Only cluster members that went
through the PossibleFalseNegative FLOSS fallback contribute strings (many
won't - that's expected, not a bug); a cluster with no contributing members
falls back to a filesize-range-only skeleton rule explicitly flagged as
needing manual work rather than being silently skipped. Written to
config.ReportDirectory/generated_rules for manual review - never
auto-imported into BinSifter's own YaraRules/CapaRules by this code.

Deliberate simplification versus the PowerShell version: that version
persisted each file's raw FLOSS JSON output to disk
(ReportDirectory/floss_reports/<sha1>.json) during the per-file scan pass,
then re-read and re-parsed those files back from disk during this later
clustering pass - a round-trip that made sense for its out-of-process
runspace architecture. Since this Python port already holds every file's
FLOSS static strings in memory for the duration of scan_directory() (no
separate process to hand data back from), engine.py passes them straight
through as an in-memory dict instead - no JSON files are read or written
by this module. Functionally equivalent, less I/O.

Known, deliberately-NOT-fixed quirk carried over from the original: the
YARA condition for a rule with common strings always reads "(3 of them)"
regardless of how many strings were actually found (1 to 12) - a rule
built from only 1 or 2 common strings can therefore never match anything.
Ported as-is rather than "improved" to e.g. min(3, len(strings)) of them,
since these are explicitly unreviewed drafts (the header comment already
says so) and silently changing detection logic here would defeat the
"draft to be reviewed by a human" premise as much as leaving the bug in.
Flag before changing.
"""

from __future__ import annotations

import os
import re
from dataclasses import dataclass
from datetime import datetime
from pathlib import Path

from binsifter.core.models import FileRecord

# Same constants as the PowerShell version - see module docstring.
_MIN_STRING_LEN = 8
_MAX_STRING_LEN = 128
_MAX_COMMON_STRINGS = 12
_MIN_CLUSTER_SIZE_FOR_RULE = 2
_FIXED_MATCH_COUNT = 3  # "(3 of them)" - see the known-quirk note above.

_RULE_NAME_SANITIZE_RE = re.compile(r"[^a-zA-Z0-9_]")


@dataclass
class DraftRuleGenResult:
    rules_written: int
    output_dir: str


def generate_draft_rules(
    records_by_cluster: dict[int, list[FileRecord]],
    static_strings_by_path: dict[str, list[str]],
    report_directory: str,
    ssdeep_cluster_threshold: int,
    timestamp: str | None = None,
) -> DraftRuleGenResult:
    """records_by_cluster: {cluster_id: [FileRecord, ...]} - callers should
    only include clusters with 2+ members (same as the PowerShell version's
    `if ($members.Count -lt 2) { continue }`); re-checked here too as a
    defensive guard.

    static_strings_by_path: {file_path: [raw FLOSS static strings]} - only
    present for files that actually ran FLOSS (the PossibleFalseNegative
    fallback); a cluster member missing from this dict simply doesn't
    contribute to the string intersection, same as a missing
    floss_reports/<sha1>.json did in the original.
    """
    timestamp = timestamp or datetime.now().strftime("%Y-%m-%d_%H%M%S")
    output_dir = Path(report_directory) / "generated_rules"
    output_dir.mkdir(parents=True, exist_ok=True)

    rules_written = 0

    for cluster_id, members in records_by_cluster.items():
        if len(members) < _MIN_CLUSTER_SIZE_FOR_RULE:
            continue

        common_strings = _intersect_static_strings(members, static_strings_by_path)
        size_condition = _build_size_condition(members)
        rule_name = _RULE_NAME_SANITIZE_RE.sub("_", f"bsifter_ssdeep_cluster_{cluster_id}_{timestamp}")

        lines = _build_rule_lines(
            rule_name=rule_name,
            cluster_id=cluster_id,
            cluster_size=len(members),
            threshold=ssdeep_cluster_threshold,
            common_strings=common_strings,
            size_condition=size_condition,
            timestamp=timestamp,
        )

        rule_path = output_dir / f"{rule_name}.yar"
        rule_path.write_text("\n".join(lines) + "\n", encoding="utf-8")
        rules_written += 1

    return DraftRuleGenResult(rules_written=rules_written, output_dir=str(output_dir))


def _intersect_static_strings(
    members: list[FileRecord], static_strings_by_path: dict[str, list[str]]
) -> list[str]:
    """Length-filters (8-128 chars) each member's static strings into a
    per-member set, then intersects across every member that had any -
    same as the PowerShell version's HashSet[string](Ordinal - i.e.
    case-sensitive) IntersectWith chain. Needs 2+ contributing members to
    produce anything; a single contributing member has nothing to
    intersect against, same as the original.

    Order/cap: the first contributing member's original string order is
    used as the iteration basis (a deterministic stand-in for .NET's
    HashSet enumeration order, which the original relied on for
    Select-Object -First 12 without a documented ordering guarantee),
    capped at 12 entries.
    """
    per_member_sets: list[set[str]] = []
    ordered_candidates: list[str] | None = None

    for member in members:
        if not member.Path:
            continue
        raw_strings = static_strings_by_path.get(member.Path)
        if not raw_strings:
            continue
        filtered = [s for s in raw_strings if s and _MIN_STRING_LEN <= len(s) <= _MAX_STRING_LEN]
        if not filtered:
            continue
        if ordered_candidates is None:
            ordered_candidates = filtered
        per_member_sets.append(set(filtered))

    if len(per_member_sets) < 2 or ordered_candidates is None:
        return []

    common = set(per_member_sets[0])
    for s in per_member_sets[1:]:
        common &= s

    return [s for s in ordered_candidates if s in common][:_MAX_COMMON_STRINGS]


def _build_size_condition(members: list[FileRecord]) -> str | None:
    sizes = []
    for member in members:
        if not member.Path:
            continue
        try:
            sizes.append(os.path.getsize(member.Path))
        except OSError:
            continue

    if not sizes:
        return None

    min_size = min(sizes)
    max_size = max(sizes)
    # PowerShell's `if ($minSize -and $maxSize)` treats 0 as falsy, same as
    # Python's `if min_size and max_size` - an all-empty-files edge case
    # (min_size == 0) intentionally skips the size condition, matching the
    # original rather than "fixing" it to handle a zero-byte minimum.
    if not (min_size and max_size):
        return None

    return f"filesize >= {max(0, min_size - 4096)} and filesize <= {max_size + 4096}"


def _build_rule_lines(
    *,
    rule_name: str,
    cluster_id: int,
    cluster_size: int,
    threshold: int,
    common_strings: list[str],
    size_condition: str | None,
    timestamp: str,
) -> list[str]:
    lines = [
        "// AUTO-GENERATED DRAFT - review before use. BinSifter Python rewrite.",
        f"// Built from SSDEEP cluster {cluster_id} ({cluster_size} files, threshold {threshold}).",
        f"// Common-string basis: {len(common_strings)} string(s) shared across FLOSS-analyzed cluster members.",
        f"rule {rule_name}",
        "{",
        "    meta:",
        '        source = "BinSifter auto-generated - DRAFT, not reviewed"',
        f"        cluster_size = {cluster_size}",
        f'        generated = "{timestamp}"',
    ]

    if common_strings:
        lines.append("    strings:")
        for i, s in enumerate(common_strings):
            escaped = s.replace("\\", "\\\\").replace('"', '\\"')
            lines.append(f'        $s{i} = "{escaped}"')
        lines.append("    condition:")
        cond = "uint16(0) == 0x5A4D and " + (f"{size_condition} and " if size_condition else "")
        cond += f"({_FIXED_MATCH_COUNT} of them)"
        lines.append(f"        {cond}")
    else:
        # No shared strings found (no cluster member had a usable FLOSS
        # static-strings contribution, or nothing survived the
        # intersection) - fall back to a filesize-range-only skeleton
        # that's explicitly flagged as needing manual work rather than
        # silently omitting the rule.
        lines.append("    condition:")
        fallback_cond = "uint16(0) == 0x5A4D" + (f" and {size_condition}" if size_condition else "")
        lines.append(f"        {fallback_cond} // TODO: no common strings found - add real detection logic before use")

    lines.append("}")
    return lines

File: ingot/crates/ingot-core/src/yara_rule_gen.rs
//! Draft YARA rule auto-generation per SSDEEP cluster - port of
//! `binsifter.core.yara_rule_gen`.
//!
//! Best-effort, clearly-labelled-as-draft rules built from strings common
//! to every member of a size>=2 SSDEEP cluster. Only cluster members that
//! went through the `PossibleFalseNegative` FLOSS fallback contribute
//! strings (`static_strings_by_path`); a cluster with no contributing
//! members falls back to a filesize-range-only skeleton explicitly flagged
//! as needing manual work. Written to `<report_dir>/generated_rules/` for
//! manual review - never auto-imported.
//!
//! FLOSS isn't ported yet, so `static_strings_by_path` is currently always
//! empty and every generated rule takes the skeleton fallback path - the
//! same behaviour Winnow has today. The full intersection logic is kept so
//! Phase 5 only needs to populate the map.
//!
//! Two quirks are carried over verbatim, not "fixed":
//!
//! * the condition always reads `(3 of them)` regardless of how many
//!   strings were found (a rule built from 1-2 strings can never match) -
//!   these are explicitly unreviewed drafts.
//! * an all-empty-files cluster (`min_size == 0`) skips the size condition.

use std::collections::{BTreeMap, HashMap};
use std::path::PathBuf;

use crate::model::FileRecord;

const MIN_STRING_LEN: usize = 8;
const MAX_STRING_LEN: usize = 128;
const MAX_COMMON_STRINGS: usize = 12;
const MIN_CLUSTER_SIZE_FOR_RULE: usize = 2;
const FIXED_MATCH_COUNT: u32 = 3;

#[derive(Debug, Clone)]
pub struct DraftRuleGenResult {
    pub rules_written: usize,
    pub output_dir: String,
}

fn sanitize_rule_name(s: &str) -> String {
    s.chars()
        .map(|c| {
            if c.is_ascii_alphanumeric() || c == '_' {
                c
            } else {
                '_'
            }
        })
        .collect()
}

/// `records_by_cluster`: `cluster_id -> members` (callers pass only size>=2
/// clusters; re-checked defensively). Returns how many `.yar` files were
/// written.
pub fn generate_draft_rules(
    records_by_cluster: &BTreeMap<i32, Vec<&FileRecord>>,
    static_strings_by_path: &HashMap<String, Vec<String>>,
    report_directory: &str,
    ssdeep_cluster_threshold: u32,
    timestamp: &str,
) -> std::io::Result<DraftRuleGenResult> {
    let output_dir = PathBuf::from(report_directory).join("generated_rules");
    std::fs::create_dir_all(&output_dir)?;

    let mut rules_written = 0;
    for (&cluster_id, members) in records_by_cluster {
        if members.len() < MIN_CLUSTER_SIZE_FOR_RULE {
            continue;
        }
        let common = intersect_static_strings(members, static_strings_by_path);
        let size_condition = build_size_condition(members);
        let rule_name =
            sanitize_rule_name(&format!("bsifter_ssdeep_cluster_{cluster_id}_{timestamp}"));

        let body = build_rule_lines(
            &rule_name,
            cluster_id,
            members.len(),
            ssdeep_cluster_threshold,
            &common,
            size_condition.as_deref(),
            timestamp,
        );
        std::fs::write(output_dir.join(format!("{rule_name}.yar")), body + "\n")?;
        rules_written += 1;
    }

    Ok(DraftRuleGenResult {
        rules_written,
        output_dir: output_dir.to_string_lossy().into_owned(),
    })
}

fn intersect_static_strings(
    members: &[&FileRecord],
    static_strings_by_path: &HashMap<String, Vec<String>>,
) -> Vec<String> {
    use std::collections::HashSet;

    let mut per_member_sets: Vec<HashSet<String>> = Vec::new();
    let mut ordered_candidates: Option<Vec<String>> = None;

    for member in members {
        if member.path.is_empty() {
            continue;
        }
        let Some(raw) = static_strings_by_path.get(&member.path) else {
            continue;
        };
        let filtered: Vec<String> = raw
            .iter()
            .filter(|s| !s.is_empty() && (MIN_STRING_LEN..=MAX_STRING_LEN).contains(&s.len()))
            .cloned()
            .collect();
        if filtered.is_empty() {
            continue;
        }
        if ordered_candidates.is_none() {
            ordered_candidates = Some(filtered.clone());
        }
        per_member_sets.push(filtered.into_iter().collect());
    }

    let (Some(candidates), true) = (ordered_candidates, per_member_sets.len() >= 2) else {
        return Vec::new();
    };

    let mut common = per_member_sets[0].clone();
    for s in &per_member_sets[1..] {
        common.retain(|x| s.contains(x));
    }

    candidates
        .into_iter()
        .filter(|s| common.contains(s))
        .take(MAX_COMMON_STRINGS)
        .collect()
}

fn build_size_condition(members: &[&FileRecord]) -> Option<String> {
    let sizes: Vec<u64> = members
        .iter()
        .filter(|m| !m.path.is_empty())
        .filter_map(|m| std::fs::metadata(&m.path).ok().map(|md| md.len()))
        .collect();
    let (min_size, max_size) = (*sizes.iter().min()?, *sizes.iter().max()?);
    if min_size == 0 || max_size == 0 {
        return None;
    }
    Some(format!(
        "filesize >= {} and filesize <= {}",
        min_size.saturating_sub(4096),
        max_size + 4096
    ))
}

#[allow(clippy::too_many_arguments)]
fn build_rule_lines(
    rule_name: &str,
    cluster_id: i32,
    cluster_size: usize,
    threshold: u32,
    common_strings: &[String],
    size_condition: Option<&str>,
    timestamp: &str,
) -> String {
    let mut lines = vec![
        "// AUTO-GENERATED DRAFT - review before use. BinSifter Ingot (Rust variant).".to_string(),
        format!("// Built from SSDEEP cluster {cluster_id} ({cluster_size} files, threshold {threshold})."),
        format!(
            "// Common-string basis: {} string(s) shared across FLOSS-analyzed cluster members.",
            common_strings.len()
        ),
        format!("rule {rule_name}"),
        "{".to_string(),
        "    meta:".to_string(),
        r#"        source = "BinSifter auto-generated - DRAFT, not reviewed""#.to_string(),
        format!("        cluster_size = {cluster_size}"),
        format!(r#"        generated = "{timestamp}""#),
    ];

    if !common_strings.is_empty() {
        lines.push("    strings:".to_string());
        for (i, s) in common_strings.iter().enumerate() {
            let escaped = s.replace('\\', "\\\\").replace('"', "\\\"");
            lines.push(format!(r#"        $s{i} = "{escaped}""#));
        }
        lines.push("    condition:".to_string());
        let mut cond = "uint16(0) == 0x5A4D and ".to_string();
        if let Some(sc) = size_condition {
            cond.push_str(sc);
            cond.push_str(" and ");
        }
        cond.push_str(&format!("({FIXED_MATCH_COUNT} of them)"));
        lines.push(format!("        {cond}"));
    } else {
        lines.push("    condition:".to_string());
        let mut cond = "uint16(0) == 0x5A4D".to_string();
        if let Some(sc) = size_condition {
            cond.push_str(&format!(" and {sc}"));
        }
        lines.push(format!(
            "        {cond} // TODO: no common strings found - add real detection logic before use"
        ));
    }

    lines.push("}".to_string());
    lines.join("\n")
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::path::Path;

    fn rec(path: &str) -> FileRecord {
        FileRecord::new(path)
    }

    #[test]
    fn skeleton_rule_when_no_strings() {
        let dir = tempfile::tempdir().unwrap();
        let a = dir.path().join("a.bin");
        let b = dir.path().join("b.bin");
        std::fs::write(&a, vec![0u8; 5000]).unwrap();
        std::fs::write(&b, vec![0u8; 9000]).unwrap();

        let (ra, rb) = (rec(&a.to_string_lossy()), rec(&b.to_string_lossy()));
        let mut by_cluster = BTreeMap::new();
        by_cluster.insert(0, vec![&ra, &rb]);

        let res = generate_draft_rules(
            &by_cluster,
            &HashMap::new(),
            &dir.path().to_string_lossy(),
            40,
            "2026-09-10_120000",
        )
        .unwrap();
        assert_eq!(res.rules_written, 1);

        // the timestamp's dashes are sanitised to underscores in the filename
        // and rule name, same as the Python/PowerShell versions
        let out = std::fs::read_to_string(
            Path::new(&res.output_dir).join("bsifter_ssdeep_cluster_0_2026_09_10_120000.yar"),
        )
        .unwrap();
        assert!(out.contains("rule bsifter_ssdeep_cluster_0_2026_09_10_120000"));
        assert!(out.contains("AUTO-GENERATED DRAFT"));
        assert!(out.contains("uint16(0) == 0x5A4D and filesize >= 904 and filesize <= 13096"));
        assert!(out.contains("TODO: no common strings"));
        assert!(!out.contains("strings:"));
    }

    #[test]
    fn strings_rule_from_common_intersection() {
        let dir = tempfile::tempdir().unwrap();
        let a = dir.path().join("a.bin");
        let b = dir.path().join("b.bin");
        std::fs::write(&a, vec![1u8; 100]).unwrap();
        std::fs::write(&b, vec![1u8; 100]).unwrap();

        let mut strings = HashMap::new();
        strings.insert(
            a.to_string_lossy().into_owned(),
            vec![
                "shared_marker_one".to_string(),
                "only_in_a_here".to_string(),
                "shared_marker_two".to_string(),
            ],
        );
        strings.insert(
            b.to_string_lossy().into_owned(),
            vec![
                "shared_marker_two".to_string(),
                "only_in_b_zzz".to_string(),
                "shared_marker_one".to_string(),
            ],
        );

        let (ra, rb) = (rec(&a.to_string_lossy()), rec(&b.to_string_lossy()));
        let mut by_cluster = BTreeMap::new();
        by_cluster.insert(5, vec![&ra, &rb]);

        let res = generate_draft_rules(
            &by_cluster,
            &strings,
            &dir.path().to_string_lossy(),
            40,
            "T",
        )
        .unwrap();
        let out = std::fs::read_to_string(
            Path::new(&res.output_dir).join("bsifter_ssdeep_cluster_5_T.yar"),
        )
        .unwrap();
        // first contributing member's order = a's order: marker_one then marker_two
        assert!(out.contains(r#"$s0 = "shared_marker_one""#));
        assert!(out.contains(r#"$s1 = "shared_marker_two""#));
        assert!(!out.contains("only_in_a"));
        assert!(out.contains("(3 of them)"));
    }

    #[test]
    fn too_small_cluster_skipped() {
        let dir = tempfile::tempdir().unwrap();
        let solo = rec("solo.bin");
        let mut by_cluster = BTreeMap::new();
        by_cluster.insert(0, vec![&solo]);
        let res = generate_draft_rules(
            &by_cluster,
            &HashMap::new(),
            &dir.path().to_string_lossy(),
            40,
            "T",
        )
        .unwrap();
        assert_eq!(res.rules_written, 0);
    }
}
